### ISSUE TITLE
Possible bugs in M4 implementation

### ISSUE THREAD

**Issue by Ko-** [KULeuven-COSIC/SABER#1]
Hello Jose, Angshuman, and others,

Congrats with your CHES paper. I tried to integrate the new [M4 implementation](https://github.com/KULeuven-COSIC/SABER/tree/master/Cortex-M/Cortex-M4/src/m4-tradeoff) into [pqm4](https://github.com/mupq/pqm4/). However, at first I didn't get any output and the compiler generated several warnings, so I went ahead and took a look.

```
$ make IMPLEMENTATION_PATH=crypto_kem/saber-bak/m4 bin/crypto_kem_saber-bak_m4_test.bin
mkdir -p elf
arm-none-eabi-gcc -o elf/crypto_kem_saber-bak_m4_test.elf -O3 -Wall -Wextra -Wimplicit-function-declaration -Wredundant-decls -Wmissing-prototypes -Wstrict-prototypes -Wundef -Wshadow -I./libopencm3/include -fno-common -mthumb -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -MD -DSTM32F4 -DMUPQ_NAMESPACE= \
	mupq/crypto_kem/test.c mupq/common/fips202.c mupq/common/sha2.c mupq/common/sp800-185.c mupq/common/nistseedexpander.c common/keccakf1600.S mupq/common/aes.c common/aes.S common/randombytes.c crypto_kem/saber-bak/m4/cbd.c crypto_kem/saber-bak/m4/fips202.c crypto_kem/saber-bak/m4/kem.c crypto_kem/saber-bak/m4/pack_unpack.c crypto_kem/saber-bak/m4/poly.c crypto_kem/saber-bak/m4/poly_mul.c crypto_kem/saber-bak/m4/SABER_indcpa.c crypto_kem/saber-bak/m4/verify.c crypto_kem/saber-bak/m4/mult_toom4_256_16.s crypto_kem/saber-bak/m4/asm_funcs.S common/hal-stm32f4.c \
	-Icrypto_kem/saber-bak/m4 -I"mupq/common" -I"common" --static -Wl,--start-group -lc -lgcc -lnosys -Wl,--end-group -Tldscripts/stm32f405x6.ld -nostartfiles -Wl,--gc-sections -mthumb -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -L./libopencm3/lib -lm -lopencm3_stm32f4
crypto_kem/saber-bak/m4/cbd.c:35:6: warning: no previous prototype for 'cbd' [-Wmissing-prototypes]
   35 | void cbd(uint16_t *r, const unsigned char *buf)
      |      ^~~
crypto_kem/saber-bak/m4/fips202.c: In function 'keccak_absorb':
crypto_kem/saber-bak/m4/fips202.c:85:5: warning: implicit declaration of function 'KeccakF1600_StatePermute' [-Wimplicit-function-declaration]
   85 |     KeccakF1600_StatePermute(s);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
crypto_kem/saber-bak/m4/kem.c: In function 'crypto_kem_dec':
crypto_kem/saber-bak/m4/kem.c:70:45: warning: passing argument 4 of 'indcpa_kem_enc_cmp' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   70 |   fail = indcpa_kem_enc_cmp(buf, kr+32, pk, c);
      |                                             ^
In file included from crypto_kem/saber-bak/m4/kem.c:5:
crypto_kem/saber-bak/m4/SABER_indcpa.h:14:124: note: expected 'unsigned char *' but argument is of type 'const unsigned char *'
   14 | unsigned char indcpa_kem_enc_cmp(unsigned char *message, unsigned char *noiseseed, const unsigned char *pk, unsigned char *ciphertext);
      |                                                                                                             ~~~~~~~~~~~~~~~^~~~~~~~~~
crypto_kem/saber-bak/m4/pack_unpack.c: In function 'POLp2BS':
crypto_kem/saber-bak/m4/pack_unpack.c:261:11: warning: unused variable 'i' [-Wunused-variable]
  261 |  uint32_t i,j;
      |           ^
crypto_kem/saber-bak/m4/pack_unpack.c: In function 'POLp2BS_cmp':
crypto_kem/saber-bak/m4/pack_unpack.c:305:11: warning: unused variable 'i' [-Wunused-variable]
  305 |  uint32_t i,j;
      |           ^
crypto_kem/saber-bak/m4/SABER_indcpa.c:32:6: warning: no previous prototype for 'byte_bank2pol_part' [-Wmissing-prototypes]
   32 | void byte_bank2pol_part(unsigned char *bytes, uint16_t pol_part[], uint16_t pol_part_start_index, uint16_t num_8coeff)
      |      ^~~~~~~~~~~~~~~~~~
crypto_kem/saber-bak/m4/SABER_indcpa.c:54:6: warning: no previous prototype for 'GenMatrix' [-Wmissing-prototypes]
   54 | void GenMatrix(polyvec *a, const unsigned char *seed)
      |      ^~~~~~~~~
crypto_kem/saber-bak/m4/SABER_indcpa.c:123:6: warning: no previous prototype for 'GenMatrix_poly' [-Wmissing-prototypes]
  123 | void GenMatrix_poly(uint16_t temp_ar[], const unsigned char *seed, uint16_t poly_number)
      |      ^~~~~~~~~~~~~~
crypto_kem/saber-bak/m4/SABER_indcpa.c: In function 'indcpa_kem_dec':
crypto_kem/saber-bak/m4/SABER_indcpa.c:401:13: warning: unused variable 'j' [-Wunused-variable]
  401 |  uint32_t i,j;
      |             ^
crypto_kem/saber-bak/m4/SABER_indcpa.c: In function 'MatrixVectorMul':
crypto_kem/saber-bak/m4/SABER_indcpa.c:483:108: warning: unused parameter 'mod' [-Wunused-parameter]
  483 | void MatrixVectorMul(polyvec *a, uint16_t skpv[SABER_K][SABER_N], uint16_t res[SABER_K][SABER_N], uint16_t mod, int16_t transpose){
      |                                                                                                   ~~~~~~~~~^~~
crypto_kem/saber-bak/m4/SABER_indcpa.c: In function 'InnerProd':
crypto_kem/saber-bak/m4/SABER_indcpa.c:554:5: warning: "/*" within comment [-Wcomment]
  554 |    }/**/
      |      
crypto_kem/saber-bak/m4/SABER_indcpa.c:537:89: warning: unused parameter 'mod' [-Wunused-parameter]
  537 | void InnerProd(uint16_t pkcl[SABER_K][SABER_N],uint16_t skpv[SABER_K][SABER_N],uint16_t mod,uint16_t res[SABER_N])
      |                                                                                ~~~~~~~~~^~~
crypto_kem/saber-bak/m4/SABER_indcpa.c: In function 'MatrixVectorMul_keypair':
crypto_kem/saber-bak/m4/SABER_indcpa.c:560:14: warning: unused variable 'k' [-Wunused-variable]
  560 |  int32_t i,j,k;
      |              ^
crypto_kem/saber-bak/m4/SABER_indcpa.c:558:131: warning: unused parameter 'mod' [-Wunused-parameter]
  558 | void MatrixVectorMul_keypair(const unsigned char *seed, uint16_t skpv[SABER_K][SABER_N], uint16_t res[SABER_K][SABER_N], uint16_t mod)
      |                                                                                                                          ~~~~~~~~~^~~
crypto_kem/saber-bak/m4/SABER_indcpa.c: In function 'MatrixVectorMul_encryption':
crypto_kem/saber-bak/m4/SABER_indcpa.c:571:129: warning: unused parameter 'mod' [-Wunused-parameter]
  571 | void MatrixVectorMul_encryption(const unsigned char *seed, uint16_t skpv[SABER_K][SABER_N], unsigned char *ciphertext, uint16_t mod)
      |                                                                                                                        ~~~~~~~~~^~~
crypto_kem/saber-bak/m4/SABER_indcpa.c: At top level:
crypto_kem/saber-bak/m4/SABER_indcpa.c:614:15: warning: no previous prototype for 'MatrixVectorMul_encryption_cmp' [-Wmissing-prototypes]
  614 | unsigned char MatrixVectorMul_encryption_cmp(const unsigned char *seed, uint16_t skpv[SABER_K][SABER_N], unsigned char *ciphertext, uint16_t mod)
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
crypto_kem/saber-bak/m4/SABER_indcpa.c: In function 'MatrixVectorMul_encryption_cmp':
crypto_kem/saber-bak/m4/SABER_indcpa.c:614:142: warning: unused parameter 'mod' [-Wunused-parameter]
  614 | unsigned char MatrixVectorMul_encryption_cmp(const unsigned char *seed, uint16_t skpv[SABER_K][SABER_N], unsigned char *ciphertext, uint16_t mod)
      |                                                                                                                                     ~~~~~~~~~^~~
crypto_kem/saber-bak/m4/SABER_indcpa.c: In function 'indcpa_kem_enc_cmp':
crypto_kem/saber-bak/m4/SABER_indcpa.c:683:7: warning: right shift count >= width of type [-Wshift-count-overflow]
  683 |  fail >>= 63;
      |       ^~~
crypto_kem/saber-bak/m4/verify.c:12:5: warning: no previous prototype for 'verify' [-Wmissing-prototypes]
   12 | int verify(const unsigned char *a, const unsigned char *b, size_t len)
      |     ^~~~~~
crypto_kem/saber-bak/m4/verify.c:26:6: warning: no previous prototype for 'cmov' [-Wmissing-prototypes]
   26 | void cmov(unsigned char *r, const unsigned char *x, size_t len, unsigned char b)
      |      ^~~~
mkdir -p bin
arm-none-eabi-objcopy -Obinary elf/crypto_kem_saber-bak_m4_test.elf bin/crypto_kem_saber-bak_m4_test.bin
rm elf/crypto_kem_saber-bak_m4_test.elf
```

- Most warnings were about missing function prototypes and unused parameters/variables. Easy to fix, but these could not have caused any real problems.
- `keccak_absorb`, `keccak_squeezeblocks` and the internal Keccak state are explicitly not exposed anymore in our current fips202 API, so I changed that to use a `shake128ctx` in `GenMatrix_poly` in [SABER_indcpa.c](https://github.com/KULeuven-COSIC/SABER/blob/master/Cortex-M/Cortex-M4/src/m4-tradeoff/SABER_indcpa.c#L133).
- [At the end](https://github.com/KULeuven-COSIC/SABER/blob/master/Cortex-M/Cortex-M4/src/m4-tradeoff/SABER_indcpa.c#L683) of `indcpa_kem_enc_cmp` the variable `fail` gets right-shifted by 63, but `fail` is of type `unsigned char`, so that won't work. Also, in the line before, the `+1` will already get truncated if `~fail` is 0xff. I used a `uint32_t` instead.
- In `crypto_kem_dec`, `c` is marked as const. However, in `indcpa_kem_enc_cmp`, `ciphertext` is no longer const and it looks like it even gets overwritten in `SABER_pack_4bit_cmp`. I copied the data to a new buffer to avoid this.

My update of your code is now in [this branch](https://github.com/mupq/pqm4/tree/saber-tches2020/crypto_kem/saber/m4). I do get output now, but unfortunately I still get wrong test values. The shared secret generated during encapsulation is not the same as after decapsulation. Perhaps I screwed something up somewhere, but I don't really see where. Just wondering: since you based this on our pqm4 framework, did you use our `test.py` and `testvectors.py` to verify that everything was still correct? Or did you have your own tests?

**Comment by josebmera** [comment]
Hi Ko,

I am so sorry that after I contacted you by mail (I think back in May?), this went completely out of my memory.
I was unable to reproduce the problem, but one of our colleagues, also in Saber team, told us about a similar problem that apparently he fixed just by upgrading the version of arm-none-eabi-gcc. I keep on using the same environment as in here for Saber, and it never gives me problems. What I would suggest is that you can try with a newer version of the compiler and let me know if the problem persists.
Anyway, in the light of the 3rd round of the NIST PQC effort we will likely provide new implementations for Cortex-M4 as we have done already for all high-end platforms.

Best regards,

**Comment by Ko-** [comment]
Hi, I also nearly forgot about this... I don't think my arm-none-eabi-gcc version was outdated though. I'm pretty sure I was on 9.3 back in April, the latest version that was available at the time.

**Comment by Ko-** [comment]
So I just tried again with arm-none-eabi-gcc 10.2, which is currently the latest version. I tried both [this branch of the pqm4 project](https://github.com/mupq/pqm4/tree/saber-tches2020/crypto_kem/saber/m4) (with the minor changes as described above) and [your original code in the SABER repo](https://github.com/KULeuven-COSIC/SABER/tree/master/Cortex-M_Implementation_KEM/Cortex-M4/src/m4-tradeoff). In both cases it doesn't pass `test.py` for me. `crypto_kem_saber_m4-tradeoff_test.bin` outputs `ERROR KEYS`.

**Comment by josebmera** [comment]
I just talked to my colleague and apparently I got it all wrong: the bug happens with recent versions of gcc
> def 9.0, but maybe already 8.0,

but the code works fine in older versions of gcc, so what he did was downgrading arm-none-eabi-gcc and not upgrading it as I said.
Given that the bug occurs for newer versions of the compiler rather than older, it is definitely something we are going to look at and fix. Also, thanks a lot for your input, I will close the issue once this is fixed.

**Comment by josebmera** [comment]
The bug with the CCA check is now fixed. As for the bug that was only occurring for certain compiler versions, it was pure luck whether or not it was happening: r12 is actually considered a scratch register that is not necessarily preserved by compiler generated functions. The issue was that our assembly was calling to memset at some point which would overwrite or not r12 depending on the compiler version and hence cause or not an error in the computations. This is also fixed now. Lastly, pqm4 has changed quite a bit the interface, we are working on integrating our code (and newer code we will make public) into the pqm4 but meanwhile we have removed the compatibility claim from the README too.
